### PR TITLE
Removed all per-entry allocations in Journal operations

### PR DESCRIPTION
**Pull request by ivankelly**
This change was originally b2cc158a from the yahoo-4.3 branch.

### CODE REVIEW

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/Journal.java [hunk 1]
```diff
@@ -24,20 +24,24 @@
...
-import java.util.LinkedList;
 import java.util.List;
+import java.util.concurrent.BlockingQueue;
 import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Executors;
-import java.util.concurrent.LinkedBlockingQueue;
 import java.util.concurrent.TimeUnit;
 import org.apache.bookkeeper.bookie.LedgerDirsManager.NoWritableLedgerDirException;
+
 import org.apache.bookkeeper.conf.ServerConfiguration;
```
Comment (sijie) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/Journal.java]: nit: remove empty spaces

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/Journal.java [hunk 2]
```diff
@@ -359,8 +375,48 @@ public void closeFileIfNecessary() {
...
+        req.logFile = logFile;
+        req.logId = logId;
+        req.lastFlushedPosition = lastFlushedPosition;
+        req.shouldClose = shouldClose;
+        req.isMarker = isMarker;
+        forceWriteQueueSize.inc();
+        return req;
+    }
+
+    private final Recycler<ForceWriteRequest> ForceWriteRequestsRecycler
```
Comment (sijie) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/Journal.java]: "forceWriteRequestRecycler" - lowercase "f"

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/Journal.java [hunk 3]
```diff
@@ -24,20 +24,24 @@
 import com.google.common.base.Stopwatch;
 import io.netty.buffer.ByteBuf;
 import io.netty.buffer.Unpooled;
+import io.netty.util.Recycler;
+import io.netty.util.Recycler.Handle;
+import io.netty.util.internal.RecyclableArrayList;
```
Comment (sijie) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/Journal.java]: It doesn't sound a good idea to use netty internal classes. can we avoid this? if we can avoid, I would prefer we have a similar util class. so we are not subject to netty changes.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/Journal.java [hunk 4]
```diff
@@ -335,11 +351,11 @@ public int process(boolean shouldForceWrite) throws IOException {
                 lastLogMark.setCurLogMark(this.logId, this.lastFlushedPosition);
 
                 // Notify the waiters that the force write succeeded
-                for (QueueEntry e : this.forceWriteWaiters) {
-                    cbThreadPool.execute(e);
+                for (int i = 0; i < forceWriteWaiters.size(); i++) {
+                    cbThreadPool.execute((QueueEntry) forceWriteWaiters.get(i));
```
Comment (sijie) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/Journal.java]: it is not very good for type cast. but I think this question is related to the comment above - can we avoid using netty internal class. if we did avoid, can we make it type safe.

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/Journal.java (hunk 3 above)
Comment (ivankelly) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/Journal.java]: It's a little nasty, but any time we change netty versions we have a bunch of changes to make anyhow, right @merlimat

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/Journal.java (hunk 1 above)
Comment (ivankelly) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/Journal.java]: Will do

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/Journal.java (hunk 4 above)
Comment (ivankelly) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/Journal.java]: Ya, the casts suck. I don't understand why the linter isn't complaining.

**Comment by ivankelly** [comment]
I'll look at copying in the recyclable array list and adding generics to it.

**Comment by ivankelly** [comment]
@sijie @eolivelli added our own RecycleableArrayList and added generics. Take another look.

**Comment by ivankelly** [comment]
retest

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/util/collections/RecyclableArrayList.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+/**
+ * A simple list which is recyclable.
+ */
+public final class RecyclableArrayList<T> extends ArrayList<T> {
+
+    private static final int DEFAULT_INITIAL_CAPACITY = 8;
+
+    public static class Recycler<X>
+        extends io.netty.util.Recycler<RecyclableArrayList<X>> {
```
Comment (sijie) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/util/collections/RecyclableArrayList.java]: nit: Recycler is already imported. so you can simplify this with `Handler`.

Comment (ivankelly) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/util/collections/RecyclableArrayList.java]: will update

**Pull request closed without merging**